lib/collectors.js

```diff
@@ -2,6 +2,46 @@
 
 module.exports = (api) => {
 
+  api.metasync.collect = (expected) => {
+
+    const isCount = typeof(expected) === 'number';
+    const isKeys = Array.isArray(expected);
+    if (!(isCount || isKeys)) throw new Error('Unexpected type');
+    const keys = expected;
+    if (isKeys) expected = expected.length;
+    let count = 0;
+    let timeout = 0;
+    let timer = null;
+    let done = null;
+    const data = {};
+
+    const collector = (key, value) => {
+      if (isCount || keys.includes(key)) count++;
+      if (value instanceof Error) return done(value, data);
+      data[key] = value;
+      if (expected === count) {
+        if (timer) clearTimeout(timer);
+        done(null, data);
+      }
+      return collector;
+    };
+
+    collector.timeout = (milliseconds) => {
+      timeout = milliseconds;
+      if (timeout) {
+        timer = setTimeout(() => {
+          const err = new Error('Collector timeout');
+          done(err, data);
```

Comment:
This should probably set some flag to indicate that `done` has already been invoked which will be checked in `collector()` so that the callback doesn't get fired twice.